lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1058,3 +1058,36 @@ void PartialConnectOp::getCanonicalizationPatterns(RewritePatternSet &results,
                                                    MLIRContext *context) {
   addCanonicalizerMethod<PartialConnectOp>(results, context);
 }
+
+LogicalResult MemOp::canonicalize(MemOp op, PatternRewriter &rewriter) {
+  // If memory has known, but zero width, eliminate it.  LowerTypes is run
+  // before this, so we can't leave structs sitting around.  Thus we must do the
+  // recursion to fix up the ports.
+  if (op.getDataType().getBitWidthOrSentinel() != 0)
+    return failure();
+  for (auto port : op->getResults()) {
+    for (auto user : port.getUsers()) {
```

Comment:
I don't think it is safe to iterate over the user list while deleting the users, plz use a pattern like this:

```
while (port.use_empty()) {
  SubfieldOp sfop = cast<SubfieldOp>(*port.use_begin());
```